torch/csrc/autograd/python_variable.cpp

```diff
@@ -413,6 +427,189 @@ static PyObject * THPVariable_device(THPVariable* self) {
   END_HANDLE_TH_ERRORS
 }
 
+// Returns the type object of the most derived subclass of torch._C._TensorBase.
+PyTypeObject *THPVariable_result_ptype(PyObject *self, PyObject *args) {
+    HANDLE_TH_ERRORS
+    PyTypeObject *ptype;
+    PyObject *obj;
+    
+    //Initialize on First Call
+    bool initialized = false;
+    static PyObject *base, *parameter;
+    if (!initialized){
+        base = THPVariableClass;
+        parameter = PyObject_GetAttrString(PyImport_ImportModule("torch.nn.parameter"), "Parameter");
+        if (parameter == NULL)
+            return NULL;
+        initialized = true;
+    }
+
+    ptype = (PyTypeObject*)base;
+    Py_INCREF(ptype);
+    if (self != NULL && PyObject_IsInstance(self, base) && !PyObject_IsInstance(self, parameter) && PyObject_IsInstance(self, (PyObject*)ptype)){
```

Comment:
I expected to see some sort of fast path here for the case of `Tensor` and `Parameter`. Also, I see you have special-cased Parameter here, and it would be good if it were described what exactly this special case is doing (and why we actually need to special-case Parameter in this way; it's not obvious to me that Parameter should be special-cased here.)

Reply:
For example, the Numpy `__array_ufunc__` has a special case for the case when the ufunc is identical to `ndarray.__array_ufunc__`, which I believe lets you get "Parameter" like behavior, because Parameter inherits the ufunc from Tensor, which creates Tensors (not Parameters!)

Reply:
(It is difficult for me to argue about this, because there is not a spec of the desired behavior clearly written down anywhere!)

Reply:
> "I expected to see some sort of fast path here..."

A: It's not a fast path. The simplest implementation has to look at the python type of every input argument as follows:
   1. self (The 1st argument).
   2. Then, the args tuple (The 2nd, ...., nth argument).
I have assumed simple inheritance, therefore I only analyze the ptype of each arg from left to right to find the most derived subclass of either np.NDarray OR torch.Tensor.

> "Also, I see you have special-cased Parameter here...."

I will add comments to say torch.nn.Parameter explicitly does not perform ptype propagation. This is because:
1. Parameters designate inputs to the neural network.
2. And, All other derived quantities should upcast to torch.Tensor so that they are not treated as tunable parameters in the neural network.
I return the most derived subclass of Tensor that is not a Parameter. 

Eg. Parameter + Tensor_Subclass = Tensor_Subclass.

Reply:
> "For example, the Numpy __array_ufunc__ has a special case for the case when the ufunc is identical to ndarray.__array_ufunc__, which I believe lets you get "Parameter" like behaviour"

This could arguably go two ways:
1. Parameter + Tensor_Subclass = Tensor_Subclass.  This is the desirable behaviour for me, but I need to process the python type of every input arg.
1. Parameter + Tensor_Subclass = Tensor. This would allow for a fast-track if for-example the first argument was a Parameter, but I can't see this being much faster.

Reply:
> "(It is difficult for me to argue about this, because there is not a spec of the desired behavior clearly written down anywhere!)"

Yes, I'm in conflict the implantation of Parameter. I wish the difference between a Tensor and a Parameter was a boolean flag stored inside the metadata of the tensor, but I'm sure others will disagree.

Reply:
Let's talk about this more on #22235. I don't mind Parameter too much but subclass propagation rules are a bit funny.